stan/math/prim/mat/fun/gp_exp_quad_cov.hpp

```diff
@@ -177,50 +180,60 @@ gp_exp_quad_cov(const std::vector<T_x1>& x1, const std::vector<T_x2>& x2,
  * @throw std::domain_error if sigma <= 0, l <= 0, or
  *   x is nan or infinite
  */
-template <typename T_x1, typename T_x2, typename T_sigma, typename T_l>
+template <typename T_x1, typename T_x2, typename T_s, typename T_l>
 inline typename Eigen::Matrix<
-    typename stan::return_type<T_x1, T_x2, T_sigma, T_l>::type, Eigen::Dynamic,
+    typename stan::return_type<T_x1, T_x2, T_s, T_l>::type, Eigen::Dynamic,
     Eigen::Dynamic>
-gp_exp_quad_cov(const std::vector<T_x1>& x1, const std::vector<T_x2>& x2,
-                const T_sigma& sigma, const std::vector<T_l>& length_scale) {
+gp_exp_quad_cov(const std::vector<Eigen::Matrix<T_x1, Eigen::Dynamic, 1>> &x1,
+                const std::vector<Eigen::Matrix<T_x2, Eigen::Dynamic, 1>> &x2,
+                const T_s &sigma, const std::vector<T_l> &length_scale) {
   using std::exp;
-  check_positive("gp_exp_quad_cov", "marginal variance", sigma);
-  for (size_t n = 0; n < x1.size(); ++n)
-    check_not_nan("gp_exp_quad_cov", "x1", x1[n]);
-  for (size_t n = 0; n < x2.size(); ++n)
-    check_not_nan("gp_exp_quad_cov", "x2", x2[n]);
-  for (size_t n = 0; n < length_scale.size(); ++n)
-    check_positive("gp_exp_quad_cov", "length-scale", length_scale[n]);
-  for (size_t n = 0; n < length_scale.size(); ++n)
-    check_not_nan("gp_exp_quad_cov", "length-scale", length_scale[n]);
-
-  Eigen::Matrix<typename stan::return_type<T_x1, T_x2, T_sigma, T_l>::type,
+  size_t x1_size = x1.size();
+  size_t x2_size = x2.size();
+  size_t l_size = length_scale.size();
+
+  const char *function_name = "gp_exp_quad_cov";
+  for (size_t i = 0; i < x1_size; ++i)
+    check_not_nan(function_name, "x1", x1[i]);
+  for (size_t i = 0; i < x2_size; ++i)
+    check_not_nan(function_name, "x2", x2[i]);
+  check_positive_finite(function_name, "magnitude", sigma);
+  check_positive_finite(function_name, "length scale", length_scale);
+  check_size_match(function_name, "x dimension", x1[0].size(),
+                   "number of length scales", l_size);
+  check_size_match(function_name, "x dimension", x2[0].size(),
+                   "number of length scales", l_size);
+
+  Eigen::Matrix<typename stan::return_type<T_x1, T_x2, T_s, T_l>::type,
                 Eigen::Dynamic, Eigen::Dynamic>
-      cov(x1.size(), x2.size());
-  if (x1.size() == 0 || x2.size() == 0)
+      cov(x1_size, x2_size);
+  if (x1_size == 0 || x2_size == 0)
     return cov;
 
-  T_sigma sigma_sq = square(sigma);
-  T_l temp_exp;
+  T_s sigma_sq = square(sigma);
 
-  for (size_t i = 0; i < x1.size(); ++i) {
-    for (size_t j = 0; j < x2.size(); ++j) {
-      temp_exp = 0;
-      for (size_t k = 0; k < length_scale.size(); ++k) {
-        temp_exp += squared_distance(x1[i], x2[j]) / square(length_scale[k]);
-      }
-      cov(i, j) = sigma_sq * exp(-0.5 * temp_exp);
+  std::vector<Eigen::Matrix<typename return_type<T_x1, T_l, T_s>::type,
+                            Eigen::Dynamic, 1>>
+      x1_new = divide_columns(x1, length_scale);
```

Comment:
If I understand the logic, `length_scale` can now be:
`std::vector<double>` or `std::vector<Eigen::Matrix<T, -1, 1>>`, correct? 

What happens when someone passes in `std::vector<std::vector<T>>`?

Reply:
yeah, exactly correct. Good catch, it doesn't compile. However, earlier we (meaning you and Bob, not me) had a discussion about what to support, and we agreed to support vectors of column vectors, for a couple of reasons: to reduce amount of tests cases, vectors of column vectors is how it's described in the literature. Also, in stan language, if we're declaring as `vector[D] x[N];` this will not use c++'s `std::vector<std::vector<T>>`, am I correct? I've tested this out in stan models locally, and it works w/ cmdstan 2.17.1